### PR TITLE
Add integer overflow checking for simd_div and simd_rem

### LINKED ISSUE #1970: `simd_div` and `simd_rem` don't check for overflow cases
Both `simd_div` and `simd_rem` are missing an overflow check that is done for the regular division operators (i.e., `/` and `%`). Consider this example:

```rust
#![feature(repr_simd, platform_intrinsics)]

#[repr(simd)]
#[allow(non_camel_case_types)]
#[derive(Clone, Copy, PartialEq, Eq)]
pub struct i32x2(i32, i32);

extern "platform-intrinsic" {
    fn simd_div<T>(x: T, y: T) -> T;
    fn simd_rem<T>(x: T, y: T) -> T;
}

#[kani::proof]
fn test_normal() {
    let val_1: i32 = kani::any();
    let val_2: i32 = kani::any();
    kani::assume(val_1 == i32::MIN);
    kani::assume(val_2 == -1);
    let result: i32 = val_1 / val_2;
}

#[kani::proof]
fn test_simd() {
    let val_1 = kani::any();
    kani::assume(val_1 == i32::MIN);
    let values_1 = i32x2(val_1, val_1);
    let val_2 = kani::any();
    kani::assume(val_2 == -1);
    let values_2 = i32x2(val_2, val_2);
    let result = unsafe { simd_div(values_1, values_2) };
}
```

This is verification output for `test_normal`:

```
RESULTS:
Check 1: test_normal.assertion.1
         - Status: SUCCESS
         - Description: "attempt to divide by zero"
         - Location: tests/kani/Intrinsics/SIMD/Operators/div_rem.rs:19:23 in function test_normal

Check 2: test_normal.assertion.2
         - Status: FAILURE
         - Description: "attempt to divide with overflow"
         - Location: tests/kani/Intrinsics/SIMD/Operators/div_rem.rs:19:23 in function test_normal

Check 3: test_normal.division-by-zero.1
         - Status: SUCCESS
         - Description: "division by zero"
         - Location: tests/kani/Intrinsics/SIMD/Operators/div_rem.rs:19:23 in function test_normal
```
This is the verification output for `test_simd`:

```
Check 1: division-by-zero.1
         - Status: SUCCESS
         - Description: "division by zero"

Check 2: division-by-zero.2
         - Status: SUCCESS
         - Description: "division by zero"
```
There are several concerns here:
 * The overflow check is missing when using the vector operator.
 * The "attempt to divide by zero" check is missing when using the vector operator.

Note that Kani tweaks the output a little bit in this example. In particular, for properties with class `division-by-zero` it omits the variables, the original outputs are:
```
Check 3: test_normal.division-by-zero.1
         - Status: SUCCESS
         - Description: "division by zero in var_10 / var_11"
         - Location: tests/kani/Intrinsics/SIMD/Operators/div_rem.rs:19:23 in function test_normal
```
```
Check 1: division-by-zero.1
         - Status: SUCCESS
         - Description: "division by zero in var_16[0] / var_17[0]"

Check 2: division-by-zero.2
         - Status: SUCCESS
         - Description: "division by zero in var_16[1] / var_17[1]"
```

In https://github.com/diffblue/cbmc/issues/7367 I asked the CBMC folks about this missing check, but it turns out these checks are being generated by MIR! I don't think that the "attempt to divide by zero" property is needed here (as it's basically a duplicate of "division by zero"), but it's unclear how to insert the "attempt to divide with overflow" property for SIMD intrinsics.

### PR DESCRIPTION
**Pull request by reisnera**
### Description of changes: 

As described in #1970, the simd_div and simd_rem intrinsics lack overflow checking. This PR implements them. It also makes a minor (somewhat) related change of moving the check to before the expression in the returned `Stmt` of `codegen_simd_op_with_overflow`.

### Resolved issues:

Resolves #1970 

### Related RFC:

n/a

### Testing:

* How is this change tested?
Added new test to ensure that attempted overflow div or rem fails the check.

* Is this a refactor change?
No

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix
(Except for the minor change in `codegen_simd_op_with_overflow`)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 1]
```diff
@@ -1522,6 +1519,37 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        loc: Location,
+    ) -> Stmt {
+        let op_fun = match intrinsic {
+            "simd_div" => Expr::div,
+            "simd_rem" => Expr::rem,
+            _ => unreachable!("expected simd_div or simd_rem"),
+        };
+        let base_type = fargs[0].typ().base_type().unwrap().clone();
+        if base_type.is_signed(self.symbol_table.machine_model()) {
+            let min_int_expr = base_type.min_int_expr(self.symbol_table.machine_model());
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1538]: What happens if the base type is float?

File: tests/kani/Intrinsics/SIMD/Operators/div_overflow_fail.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// Checks that the `simd_div` and `simd_rem` intrinsics check for overflows.
+// kani-verify-fail
```
Comment (celinval) [inline tests/kani/Intrinsics/SIMD/Operators/div_overflow_fail.rs:5]: Do you mind moving this test to the `expected` suite please? I prefer using that suite so you can pin down the exact failure.  `kani-verify-fail` is too broad, and I've seen cases where the test was passing due to an unrelated failure.

I think even if kani-compiler crashes will be considered OK.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 3]
```diff
@@ -1522,6 +1519,37 @@ impl<'tcx> GotocCtx<'tcx> {
         self.codegen_expr_to_place(p, e)
     }
 
+    /// Codegen for `simd_div` and `simd_rem` intrinsics including overflow checks.
```
Comment (celinval) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1522]: can you please document which overflow you are covering in this case?

**Review by celinval — COMMENTED** [review]
Thanks for doing this! Can you please add a test for float vector too? I'm worried that this could trigger similar failure to https://github.com/model-checking/kani/issues/2631.

File: tests/kani/Intrinsics/SIMD/Operators/div_overflow_fail.rs (hunk 2 above)
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/div_overflow_fail.rs:5]: Absolutely! This makes perfect sense, thanks

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs (hunk 1 above)
Comment (reisnera) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:1538]: Thanks for pointing this out. I had my head in the sand from working on simd shifts so I apparently forgot floats existed. I added details to issue #1970.

**Comment by celinval** [comment]
Hi @reisnera, is this ready for review?

**Comment by reisnera** [comment]
Hi @celinval, yes I believe so. I made a comment regarding `simd_rem` not checking for overflows. Did you see my comment [above](https://github.com/model-checking/kani/pull/2645#issuecomment-1660933818)?

**Comment by adpaco-aws** [comment]
Hi @reisnera , thanks for your work on these intrinsics!

When checking if two floating point numbers are the same, we often allow a small amount (usually called "epsilon") to account for precision errors. For example, we use this `fp_equals` function in some of our tests (instead of `==`):

```rs
fn fp_equals(value: f32, expected: f32) -> bool {
    let abs_diff = (value - expected).abs();
    abs_diff <= f32::EPSILON
}
```

Can you try that for the floating point tests? I can point you to specific tests that make use of it if you can't find them. You may also need to add assumptions that use the `is_finite()` or `is_nan()` methods.

Regarding the missing checks, we could ask CBMC if they're interested in adding them first. That way, both tools could benefit from the check. What do you think?

**Comment by reisnera** [comment]
Hi @adpaco-aws, thanks! I've enjoyed working on these!

For `fp_equal`, are you referring to using this in the `division_float.rs` test file to try to fix the issue I'm having with `simd_rem`? Or just in general?

I can see where some tests use the fp_equals function and I'm aware of the difficulties with equality comparison of floats, but in the case of the `division_float.rs` test file, don't we want to test that the results of normal division and SIMD division are bit-equal? (Unless the `simd_rem` issue is indeed a similar type of issue as with the trigonometric functions).

> Regarding the missing checks, we could ask CBMC if they're interested in adding them first. That way, both tools could benefit from the check. What do you think?

I agree that this would be great if they could!

**Comment by reisnera** [comment]
@adpaco-aws So I tried your suggestion, and no luck. However this prompted me to investigate further and interestingly the problem might just be with `%`. I filed a separate issues with a non-SIMD example here: #2669

**Comment by adpaco-aws** [comment]
The issue you found means that you'll get nondeterministic values as a result of the `mod` operation, which makes clear why the comparison fails later.

But I don't that should be blocking this PR: we can put the cases which are currently failing into a `fixme` test and merge them once #2669 is resolved. What do you think @celinval ?

**Comment by celinval** [comment]
> The issue you found means that you'll get nondeterministic values as a result of the `mod` operation, which makes clear why the comparison fails later.
> 
> But I don't that should be blocking this PR: we can put the cases which are currently failing into a `fixme` test and merge them once #2669 is resolved. What do you think @celinval ?

As I mentioned in #2669, I believe this is a Kani bug (hence the "ignored warning"). We might need to model floating point remainder operation on our side.

I do agree that this shouldn't block this PR, since the issue is consistent for SIMD and non-SIMD (i.e.: this PR is not introducing any bad behavior). @reisnera can you please add a fixme test to our suite either as part of this PR or a separate PR that reproduces the issue you created and also add a similar harness to the one you added to the division that checks that simd produces the same value as the non-simd one.

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+fn test_simd_div() {
+    let dividend: f32 = kani::any::<i8>().into();
+    let divisor: f32 = kani::any::<i8>().into();
+    // Narrow down the divisor interval so the operation finishes in a short time
+    kani::assume(divisor != 0.0 && divisor.abs() < 5.0);
+    let normal_result = dividend / divisor;
+    let dividends = f32x2(dividend, dividend);
+    let divisors = f32x2(divisor, divisor);
+    let simd_result = unsafe { simd_div(dividends, divisors) };
+    assert_eq!(normal_result, simd_result.0);
```
Comment (celinval) [inline tests/kani/Intrinsics/SIMD/Operators/division_float.rs:28]: Can you please also check the second element of the simd_result? Ideally, we should have a test that have different simd elements, to ensure we are computing them independently. :)

File: tests/expected/intrinsics/simd-div-rem-overflow/main.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+extern "platform-intrinsic" {
+    fn simd_div<T>(x: T, y: T) -> T;
+    fn simd_rem<T>(x: T, y: T) -> T;
+}
+
+#[kani::proof]
+fn test_simd_div_overflow() {
+    let dividend = kani::any();
+    kani::assume(dividend == i32::MIN);
```
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/expected/intrinsics/simd-div-rem-overflow/main.rs:21]: Yes it is. Yeah that makes more sense, will make the change!

File: tests/expected/intrinsics/simd-div-rem-overflow/main.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    let dividends = i32x2(dividend, dividend);
+    let divisor = kani::any();
+    kani::assume(divisor == -1);
+    let divisors = i32x2(divisor, divisor);
+    let _quotient = unsafe { simd_div(dividends, divisors) };
```
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/expected/intrinsics/simd-div-rem-overflow/main.rs:26]: If I add that check, it will be unreachable as the division will overflow. I'm not sure how I should proceed with this one.

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+pub struct f32x2(f32, f32);
+
+extern "platform-intrinsic" {
+    fn simd_div<T>(x: T, y: T) -> T;
+    fn simd_rem<T>(x: T, y: T) -> T;
+}
+
+#[kani::proof]
+fn test_simd_div() {
+    let dividend: f32 = kani::any::<i8>().into();
```
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:20]: I did this to avoid overflows in the division/restrict the scope of the test. If the division overflows, I would need to adjust the equality test to account for overflow results, however then the harness won't terminate in a reasonable amount of time. Definitely open to suggestions though!

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs (hunk 4 above)
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:28]: I can definitely separate these to confirm that they're computing independently

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+#[kani::proof]
+fn test_simd_rem() {
+    let dividend: f32 = kani::any::<i8>().into();
+    let divisor: f32 = kani::any::<i8>().into();
+    // Narrow down the divisor interval so the operation finishes in a short time
+    kani::assume(divisor != 0.0 && divisor.abs() < 5.0);
+    let dividends = f32x2(dividend, dividend);
+    let divisors = f32x2(divisor, divisor);
+    let _simd_result = unsafe { simd_rem(dividends, divisors) };
```
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:39]: Will do! Sorry, what exactly do you want to wrap with black_box?

**Comment by reisnera** [comment]
@celinval and @adpaco-aws, I responded to the review above but FYI I'll also need some time to finish this up as I'll be working my usual job through the weekend into next week :)

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:20]: What if you restrict the value to a small interval, would that work? Something like:
```rust
let dividend = kani::any_where(|d: f32| d > LOWER_BOUND && d < UPPER_BOUND);
```
I just want to make sure these operations work with non-absolute values.

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:39]:
```rust
    std::hint::black_box(unsafe { simd_rem(dividends, divisors) });
```

File: tests/expected/intrinsics/simd-div-rem-overflow/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (celinval) [reply tests/expected/intrinsics/simd-div-rem-overflow/main.rs:26]: That should be OK... you can even add to your expectation that the assertion should be unreachable. :)

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:39]: Oh gotcha, that makes sense. Thought, I actually think I'm going to omit this function from this test file as the existence of simd_rem is already checked via other tests so I don't think this would add anything. I'll add the fixme test for simd_rem with floating point inputs though.

File: tests/kani/Intrinsics/SIMD/Operators/division_float.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (reisnera) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs:20]: Even with a narrow range of bounds, it doesn't seems to terminate in any reasonable amount of time (at all?). I came up with another solution similar to my first but with multiplying 0.5 by an i8 so that it isn't just round numbers. I'll push my changes a bit later so you can take a look!

[the remark this answers is not in the document]
Comment (celinval) [reply tests/kani/Intrinsics/SIMD/Operators/division_float.rs]: Awesome. Thanks

**Review by celinval — APPROVED** [review]
Thank you!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 9]
```diff
@@ -509,9 +509,12 @@ impl<'tcx> GotocCtx<'tcx> {
                 loc,
             ),
             "simd_and" => codegen_intrinsic_binop!(bitand),
-            // TODO: `simd_div` and `simd_rem` don't check for overflow cases.
+            // TODO: `simd_rem` doesn't check for overflow cases for floating point operands.
             // <https://github.com/model-checking/kani/issues/1970>
```
Comment (adpaco-aws) [inline kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:513]: This line can be removed.

**Review by adpaco-aws — COMMENTED** [review]
Thanks, @reisnera ! With this contribution and #1963, I think we can update the [documentation for platform intrinsic support](https://model-checking.github.io/kani/rust-feature-support/intrinsics.html#platform-intrinsics). Would you mind doing that?

**Comment by reisnera** [comment]
@adpaco-aws Absolutely! I'm actually not sure if I should leave something in the notes column for `simd_rem`. I'm still trying to wrap my head around if #2669 affects `simd_rem`. I think it does?

**Comment by adpaco-aws** [comment]
> @adpaco-aws Absolutely! I'm actually not sure if I should leave something in the notes column for `simd_rem`. I'm still trying to wrap my head around if #2669 affects `simd_rem`. I think it does?

Hey @reisnera , sorry for not getting back to you earlier.

If I'm not mistaken, we still need to open an issue in the CBMC repo to ask about the missing checks. We can note there's missing support for the operation with floating-point operand, and also the missing checks that should be coming from CBMC.

**Comment by reisnera** [comment]
@adpaco-aws I updated the docs. According to the comment [here](https://github.com/model-checking/kani/issues/2669#issuecomment-1677099829), it seems this is already doable on CBMC end unless I'm missing something. I'll work on a pull request for that issue separately.

**Review by adpaco-aws — APPROVED** [review]
Thanks @reisnera !
